slice2matlab parallel build failures creating module directories

This code is prone to fail when building Slice files for the same module in parallel:

https://github.com/zeroc-ice/ice/blob/bf991f4f7604869517f9211f18a6048239521551/cpp/src/slice2matlab/Main.cpp#L234-L243

We just hit this with CI builds:

```
../cpp/bin/slice2matlab  -I../slice -Itest/Ice/optional --output-dir test/Ice/optional  --depend test/Ice/optional/Test.ice | sed 's/\(.*: \\\)/\/\/\1/' > test/Ice/optional/.depend/Test.ice.d
../cpp/bin/slice2matlab  -I../slice -Itest/Ice/optional --output-dir test/Ice/optional/generated  test/Ice/optional/ClientPrivate.ice
../cpp/bin/slice2matlab  -I../slice -Itest/Ice/optional --output-dir test/Ice/optional/generated  test/Ice/optional/Test.ice
../cpp/bin/slice2matlab: error: cannot create directory `test/Ice/optional/generated/+Test': File exists
make: *** [Makefile:146: test/Ice/optional/generated/Test.m] Error 1
```

--- a/cpp/src/slice2java/JavaUtil.cpp
+++ b/cpp/src/slice2java/JavaUtil.cpp
@@ -701,7 +701,7 @@ Slice::JavaOutput::openClass(const string& cls, const string& prefix, const stri
             }
 
             int err = IceUtilInternal::mkdir(path, 0777);
-            // If slice2java is run concurrently, it's possible that another instance of slice2matlab has already
+            // If slice2java is run concurrently, it's possible that another instance of slice2java has already
             // created the directory.
             if (err == EEXIST && IceUtilInternal::directoryExists(path))
             {

--- a/cpp/src/slice2java/JavaUtil.cpp
+++ b/cpp/src/slice2java/JavaUtil.cpp
@@ -700,10 +700,17 @@ Slice::JavaOutput::openClass(const string& cls, const string& prefix, const stri
                 continue;
             }
 
-            if(IceUtilInternal::mkdir(path, 0777) != 0)
+            int err = IceUtilInternal::mkdir(path, 0777);
+            // If slice2java is run concurrently, it's possible that another instance of slice2matlab has already
+            // created the directory.
+            if (err == EEXIST && IceUtilInternal::directoryExists(path))
+            {
+                // Directory already exists, ignore the error
+            }
+            else if(err != 0)
             {
                 ostringstream os;
-                os << "cannot create directory `" << path << "': " << IceUtilInternal::errorToString(errno);
+                os << "cannot create directory `" << path << "': " << IceUtilInternal::errorToString(err);
                 throw FileException(__FILE__, __LINE__, os.str());
             }
             FileTracker::instance()->addDirectory(path);

--- a/cpp/src/slice2matlab/Main.cpp
+++ b/cpp/src/slice2matlab/Main.cpp
@@ -233,10 +233,17 @@ openClass(const string& abs, const string& dir, IceUtilInternal::Output& out)
         path += "+" + lookupKwd(v[i]);
         if(!IceUtilInternal::directoryExists(path))
         {
-            if(IceUtilInternal::mkdir(path, 0777) != 0)
+            int err = IceUtilInternal::mkdir(path, 0777);
+            // If slice2matlab is run concurrently, it's possible that another instance of slice2matlab has already
+            // created the directory.
+            if (err == EEXIST && IceUtilInternal::directoryExists(path))
+            {
+                // Directory already exists, ignore the error
+            }
+            else if(err != 0)
             {
                 ostringstream os;
-                os << "cannot create directory `" << path << "': " << IceUtilInternal::errorToString(errno);
+                os << "cannot create directory `" << path << "': " << IceUtilInternal::errorToString(err);
                 throw FileException(__FILE__, __LINE__, os.str());
             }
             FileTracker::instance()->addDirectory(path);

